dig_test.go

```diff
@@ -1433,6 +1379,89 @@ func TestGroups(t *testing.T) {
 		require.Error(t, err, "failed to provide")
 		assert.Contains(t, err.Error(), "flatten can be applied to slices only")
 	})
+
+	t.Run("soft value group", func(t *testing.T) {
+		type Param1 struct {
+			dig.In
+
+			Values []string `group:"foo,soft"`
+		}
+		type Result struct {
+			dig.Out
+
+			Value1 string `group:"foo"`
+			Value2 int
+		}
+
+		c := digtest.New(t)
+		c.RequireProvide(func() Result { return Result{Value1: "a", Value2: 2} })
+		c.RequireProvide(func() string { return "b" }, dig.Group("foo"))
+
+		// The only value that must be in the group is the one that's provided
+		// because it would be provided anyways by another dependency, in
+		// this case we need an int, so the first constructor is called, and
+		// this provides a string, which is the one in the group
+		c.RequireInvoke(func(p2 int, p1 Param1) {
+			assert.ElementsMatch(t, []string{"a"}, p1.Values)
+		})
+	})
+
+	t.Run("two soft group values provided by one constructor", func(t *testing.T) {
+		type param struct {
+			dig.In
+
+			Value1 []string `group:"foo,soft"`
+			Value2 []int    `group:"bar,soft"`
+			Value3 float32
+		}
+
+		type result struct {
+			dig.Out
+
+			Result1 []string `group:"foo,flatten"`
+			Result2 int      `group:"bar"`
+		}
+		c := digtest.New(t)
+
+		c.RequireProvide(func() result {
+			return result{
+				Result1: []string{"a", "b", "c"},
+				Result2: 4,
+			}
+		})
+		c.RequireProvide(func() float32 { return 3.1416 })
+
+		c.RequireInvoke(func(p param) {
+			assert.ElementsMatch(t, []string{}, p.Value1)
+			assert.ElementsMatch(t, []int{}, p.Value2)
+			assert.Equal(t, float32(3.1416), p.Value3)
+		})
+	})
+
+	t.Run("value group provided after a hard dependency is provided", func(t *testing.T) {
+		type Param struct {
+			dig.In
+
+			Value []string `group:"foo,soft"`
+		}
+
+		type Result struct {
+			dig.Out
+
+			Value1 string `group:"foo"`
+		}
+
+		c := digtest.New(t)
+		c.Provide(func() (Result, int) { return Result{Value1: "a"}, 2 })
+
+		c.RequireInvoke(func(param Param) {
+			assert.ElementsMatch(t, []string{}, param.Value)
+		})
+		c.RequireInvoke(func(int) {})
+		c.RequireInvoke(func(param Param) {
+			assert.ElementsMatch(t, []string{"a"}, param.Value)
+		})
+	})
```

Comment:
Let's add a test case for the canonical example of a constructor that provides 2 types:
1. a type that is only consumed by a soft value group
2. a type that is not consumed by anything

And add a `assert.Error(t, "this constructor should not be called")` in that constructor to make sure it's not getting called.